CMakeLists.txt

```diff
@@ -352,6 +366,12 @@ if(FLATBUFFERS_BUILD_SHAREDLIB)
   set_target_properties(flatbuffers_shared PROPERTIES OUTPUT_NAME flatbuffers
                         SOVERSION "${FlatBuffers_Library_SONAME_MAJOR}"
                         VERSION "${FlatBuffers_Library_SONAME_FULL}")
+  if(FLATBUFFERS_ENABLE_PCH AND COMMAND target_precompile_headers)
```

Comment:
since this is repeated, should we stick this in a function/macro?

Reply:
or in a private variable ? 
`set(FLATBUFFERS_USE_PCH FLATBUFFERS_ENABLE_PCH AND COMMAND target_precompile_headers)`

* FLATBUFFERS_USE_PCH ?
* FLATBUFFERS_PCH_AVAILABLE ?
* _FLATBUFFERS_ENABLE_PCH ?

I don't really see how to return a value from a macro/function ? Any link to documentation?

Reply:
I would recommend adding a function like `add_fsanitize_to_target`: `function(add_pch_to_target _target _pch_header)`.
It could work if `util.cpp` wasn't an exceptional case.
What problem was with this file?

Reply:
Problem is this block of code : 
``` cpp
// clang-format off
// Dont't remove `format off`, it prevent reordering of win-includes.
#define _POSIX_C_SOURCE 200112L // For stat from stat/stat.h and fseeko() (POSIX extensions).
#ifdef _WIN32
#  ifndef WIN32_LEAN_AND_MEAN
#    define WIN32_LEAN_AND_MEAN
#  endif
#  ifndef NOMINMAX
#    define NOMINMAX
#  endif
#  ifdef _MSC_VER
#    include <crtdbg.h>
#  endif
#  include <windows.h>  // Must be included before <direct.h>
#  include <direct.h>
#  include <winbase.h>
#  undef interface  // This is also important because of reasons
#else
#  define _XOPEN_SOURCE 600 // For PATH_MAX from limits.h (SUSv2 extension) 
#  include <limits.h>
#endif
// clang-format on

#include "flatbuffers/base.h"
#include "flatbuffers/util.h"

#include <sys/stat.h>
#include <clocale>
#include <fstream>

namespace flatbuffers {
```

If using precompiled header for that file, `_POSIX_C_SOURCE ` and `_XOPEN_SOURCE` are already defined.
Possible workaround would be to `#undef _POSIX_C_SOURCE` & `#undef _XOPEN_SOURCE`. Which doesn't make sense because standard library is already included at that point due to inclusion in `pch.h`.

Other workaround is to move this block of code to `pch.h`. But then this code need to be maintained in two files.

Anyway i see what the function approched is. The following code can be called safely as many time as we want. 

``` cmake
if(NOT MSVC)
  set_source_files_properties(src/util.cpp PROPERTIES SKIP_PRECOMPILE_HEADERS ON)
endif()
```

So it can be safely put inside the `add_pch_to_target` function. For me it's the best solution because it doesn't modify source code nor maintainability.

``` cmake 
function(add_pch_to_target _target _pch_header)
  if(COMMAND target_precompile_headers)
    target_precompile_headers(${_target} PRIVATE ${_pch_header})
    if(NOT MSVC)
      set_source_files_properties(src/util.cpp PROPERTIES SKIP_PRECOMPILE_HEADERS ON)
    endif()
  endif()
endfunction()
```